### PR TITLE
Gh18838 factory devfile2

### PR DESCRIPTION
**Pull request by sparkoo**
### What does this PR do?
Adds support for devfile v2 to /factory/resolve. Validation and property overrides are not working at this point.

### Screenshot/screencast of this PR

```
$ curl -k 'https://che-che.192.168.39.5.nip.io/api/factory/resolver' -H 'Content-Type: application/json' --data-raw '{"url":"https://github.com/sparkoo/spring-petclinic"}' | jq

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   534  100   481  100    53   8745    963 --:--:-- --:--:-- --:--:--  9709
{
  "v": "4.0",
  "source": "devfile.yaml",
  "devfile": {
    "schemaVersion": "2.0.0",
    "metadata": {
      "name": "spring petclinic demo"
    },
    "projects": [
      {
        "name": "spring-petclinic",
        "git": {
          "remotes": {
            "origin": "https://github.com/sparkoo/spring-petclinic"
          },
          "checkoutFrom": {
            "revision": "master",
            "remote": "origin"
          }
        }
      }
    ],
    "components": [
      {
        "name": "mysql",
        "container": {
          "image": "quay.io/eclipse/che--centos--mysql-57-centos7:latest-e08ee4d43b7356607685b69bde6335e27cf20c020f345b6c6c59400183882764",
          "memoryLimit": "128M"
        }
      }
    ]
  }
}

```

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/18838

### How to test this PR?
Install Che and call `/factory/resolve` che-server API with devfile v2. You can use my test repo https://github.com/sparkoo/spring-petclinic or any other you like with devfile v2.
Body of the request is `{"url": "<repo/devfile-url>"}` (see curl example above)

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Review by metlos — COMMENTED** [review]
`FactoryMetaDto` approach seems quite clean. 
I personally don't like the `if (x instanceof Y)` checks too much - what about introducing a visitor interface for the factories so that the callers could use that instead of a series of `instanceof` checks? But that's more of a nitpick.

`FactoryMetaDto` -> `AbstractFactoryDto`? Even though that may look like a random mix of design pattern names :smile:

**Comment by sparkoo** [comment]
@metlos Visitor implemented, looks much better imo, thanks. I've decided to return the factory from accept/visit method, so it can be chained when used.

See `FactoryVisitor`, `FactoryDto` and `FactoryDevfileV2Dto`. Then `GithubFactoryParametersResolver` with `BitbucketServerAuthorizingFactoryParametersResolver` where the Visitor is implemented and used.

File: wsmaster/che-core-api-factory-github/src/test/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolverTest.java [hunk 1]
```diff
@@ -141,7 +141,7 @@ public void shouldGenerateDevfileForFactoryWithNoDevfile() throws Exception {
         .thenReturn(Optional.empty());
     Map<String, String> params = ImmutableMap.of(URL_PARAMETER_NAME, githubUrl);
     // when
-    FactoryDto factory = githubFactoryParametersResolver.createFactory(params);
+    FactoryDto factory = (FactoryDto) githubFactoryParametersResolver.createFactory(params);
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-factory-github/src/test/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolverTest.java:144]: Cast usually is a smell of something that is not right. Can we avoid it?

**Review by skabashnyuk — COMMENTED** [review]
Looks good to me in general.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/DevfileParser.java [hunk 2]
```diff
@@ -87,36 +88,54 @@ public DevfileParser(
...
       // should never happen as we send empty overrides map
       throw new RuntimeException(e.getMessage());
     }
   }
 
   /**
-   * Creates {@link DevfileImpl} from given devfile content in YAML and provides possibility to
-   * override its values using key-value map, where key is an json-pointer-like string and value is
-   * desired property value. NOTE: unlike json pointers, objects in arrays should be pointed by
-   * their names, not by index. Examples:
```
Comment (mshaposhnik) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/DevfileParser.java:101]: imo that comment about override rules is still useful; may we keep it ?

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/DevfileVersion.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+ */
+package org.eclipse.che.api.workspace.server.devfile;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import javax.inject.Singleton;
+import org.eclipse.che.api.workspace.server.devfile.exception.DevfileException;
+
+/** Class that helps determine devfile versions. */
+@Singleton
+public class DevfileVersion {
```
Comment (mshaposhnik) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/DevfileVersion.java]: Such name is hardly giving idea that this is an utility class. Sounds more like a POJO or data object for me.
Can we add an verb, like DevfileVersionDetector\Extractor\Retriever\... ?

File: wsmaster/che-core-api-factory-github/src/test/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolverTest.java (hunk 1 above)
Comment (sparkoo) [reply wsmaster/che-core-api-factory-github/src/test/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolverTest.java:144]: For production code I agree. But this is test code. In production code, we call `createFactory` and don't further care what's inside returned data object. We only do few more steps with it and return it from the service https://github.com/eclipse/che/blob/master/wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/FactoryService.java#L102 So we do not need to cast. Here in the test, we're testing what's really inside and I think cast in the test is ok.

WDYT? Still worth to try get rid of it?

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/DevfileParser.java (hunk 2 above)
Comment (sparkoo) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/DevfileParser.java:101]: added to `parseJsonNode` method

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/DevfileVersion.java (hunk 3 above)
Comment (sparkoo) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/DevfileVersion.java]: `DevfileVersionDetector` now

**Review by mshaposhnik — APPROVED** [review]

File: wsmaster/che-core-api-factory-github/src/test/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolverTest.java (hunk 1 above)
Comment (skabashnyuk) [reply wsmaster/che-core-api-factory-github/src/test/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolverTest.java:144]: ok

**Review by skabashnyuk — APPROVED** [review]

**Review by metlos — APPROVED** [review]

**Pull request merged**